python/ray/air/execution/resources/placement_group.py

```diff
@@ -0,0 +1,213 @@
+            )
+        return annotated
+
+
+@DeveloperAPI
+class PlacementGroupResourceManager(ResourceManager):
+    """Resource manager using placement groups as the resource backend.
+
+    Ray core does not emit events when resources are available. Instead, we
+    have to periodically request and update the scheduling state of the
+    placement groups.
+
+    Internally, the placement group lifecycle is like this:
+    - Resources are requested with ``request_resources()``
+    - A placement group is scheduled ("staged")
+    - A ``PlacementGroup.ready()`` is scheduled ("staging future")
+    - We update the scheduling state when we need to
+        (e.g. when ``has_resources_ready()`` is called)
+    - When staging futures resolved, a placement group is moved from "staging"
+        to "ready"
+    - When a resource request is canceled, we remove a placement group from "staging".
+        If there are not staged placement groups (because they are already "ready"),
+        we remove one from "ready" instead.
+    - When a resource is acquired, the pg is removed from "ready" and moved
+        to "acquired"
+    - When a resource is freed, the pg is removed from "acquired" and destroyed
+
+    Per default, placement group scheduling state is refreshed every time when
+    resource state is inquired, but not more often than once every ``update_interval``
+    seconds. Alternatively, staging futures can be retrieved (and awaited) with
+    ``get_resource_futures()`` and state update can be force with ``update_state()``.
+
+    Args:
+        update_interval: Minimum interval in seconds between updating scheduling
+            state of placement groups.
+
+    """
+
+    _resource_cls: AllocatedResource = PlacementGroupAllocatedResource
+
+    def __init__(self, update_interval: float = 0.1):
+        self._pg_to_request: Dict[PlacementGroup, ResourceRequest] = {}
+        self._request_to_staged_pgs: Dict[
+            ResourceRequest, Set[PlacementGroup]
+        ] = defaultdict(set)
+        self._request_to_ready_pgs: Dict[
+            ResourceRequest, Set[PlacementGroup]
+        ] = defaultdict(set)
+
+        self._staging_future_to_pg: Dict[ray.ObjectRef, PlacementGroup] = dict()
+        self._pg_to_staging_future: Dict[PlacementGroup, ray.ObjectRef] = dict()
+        self._acquired_pgs: Set[PlacementGroup] = set()
+
+        self._update_interval = update_interval
+        self._last_update = time.monotonic() - self._update_interval - 1
+
+    def get_resource_futures(self) -> List[ray.ObjectRef]:
+        return list(self._staging_future_to_pg.keys())
+
+    def _maybe_update_state(self):
+        now = time.monotonic()
+        if now > self._last_update + self._update_interval:
+            self.update_state()
+
+    def update_state(self):
+        ready, not_ready = ray.wait(
+            list(self._staging_future_to_pg.keys()),
+            num_returns=len(self._staging_future_to_pg),
+            timeout=1e-6,
```

Comment:
Add a comment to explain this or link the issue about the unexpected behavior when `timeout=0`?

Reply:
Actually @krfricke, @gjoliver just tested this out and apparently `ray.wait(timeout=0)` will return immediately while `ray.get(timeout=0)` won't.